test/arrayops.jl

```diff
@@ -271,18 +271,30 @@ end
     @test typeof(Vector(3)) == Vector{Any}
     @test typeof(Vector()) == Vector{Any}
     @test typeof(Matrix{Int}(2,3)) == Matrix{Int}
-    @test typeof(Matrix{Int}()) == Matrix{Int}
     @test typeof(Matrix(2,3)) == Matrix{Any}
-    @test typeof(Matrix()) == Matrix{Any}
 
     @test size(Vector{Int}(3)) == (3,)
     @test size(Vector{Int}()) == (0,)
     @test size(Vector(3)) == (3,)
     @test size(Vector()) == (0,)
     @test size(Matrix{Int}(2,3)) == (2,3)
-    @test size(Matrix{Int}()) == (0,0)
     @test size(Matrix(2,3)) == (2,3)
-    @test size(Matrix()) == (0,0)
+
+    # TODO: will throw MethodError after 0.6 deprecations are deleted
+    dw = Base.JLOptions().depwarn
+    if dw == 2
+        @test_throws ErrorException Matrix{Int}()
+        @test_throws ErrorException Matrix()
+    elseif dw == 1
+        @test_warn "deprecated" Matrix{Int}()
+        @test_warn "deprecated" Matrix()
+    elseif dw == 0
+        @test size(Matrix{Int}()) == (0,0)
+        @test size(Matrix()) == (0,0)
+    else
+        error("unexpected depwarn value")
+    end
+    @test_throws MethodError Array{Int,3}()
```

Comment:
We could have a `@test_deprecated` macro that does this for you. Usage would be something like this:

```jl
@test_deprecated size(Matrix{Int}()) == (0,0)
```

which would expand to one of the following, depending on the value of `Base.JLOptions().depwarn`:

```jl
@test_throws ErrorException size(Matrix{Int}())
@test_warn "deprecated" size(Matrix{Int}())
@test size(Matrix{Int}()) == (0,0)
```

The main reason I wanted this complexity in here is so that I couldn't possibly forget to update the tests when we delete the deprecations for 1.0.